eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,451 @@
-  gasIdL,
-  gasStakePoolVotesL,
+import Cardano.Ledger.Binary (
+  DecCBOR (..),
+  DecShareCBOR (..),
+  EncCBOR (..),
+  decodeRecordNamed,
+  encodeListLen,
  )
+import Cardano.Ledger.Conway.Core
+import Cardano.Ledger.Conway.Governance.Procedures
 import Control.DeepSeq (NFData)
+import Control.Exception (assert)
 import Control.Monad (guard)
-import Data.Aeson (KeyValue, ToJSON (..), object, pairs, (.=))
+import Data.Aeson (KeyValue ((.=)), ToJSON (..), object, pairs)
 import Data.Default.Class (Default (..))
+import Data.Foldable (foldl')
 import Data.Functor (($>))
+import Data.Kind (Type)
 import qualified Data.Map as Map
 import Data.Map.Strict (Map)
+import Data.Maybe (isNothing)
 import qualified Data.OMap.Strict as OMap
+import Data.Sequence (Seq)
 import Data.Sequence.Strict (StrictSeq (..))
 import Data.Set (Set)
 import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import Lens.Micro
 import NoThunks.Class (NoThunks)
 
-newtype Proposals era
-  = Proposals
-      (OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
-  deriving newtype (Show, Eq)
-  deriving stock (Generic)
+data PRoot a = PRoot
+  { prRoot :: !(Maybe a)
+  , prChildren :: !(Set a)
+  }
+  deriving (Eq, Ord, Generic, NoThunks, NFData, Default)
+
+instance Show a => Show (PRoot a) where
+  show (PRoot root children) =
+    unlines
+      [ mconcat
+          [ "<R"
+          , case root of
+              Just r -> mconcat ["(", show r, ")"]
+              Nothing -> "(-)"
+          , ">"
+          ]
+      , mconcat
+          [ "<C"
+          , "{"
+          , show $ Set.size children
+          , "}"
+          , show $ Set.toList children
+          , ">"
+          ]
+      ]
+
+prRootL :: Lens' (PRoot a) (Maybe a)
+prRootL = lens prRoot $ \x y -> x {prRoot = y}
+
+prChildrenL :: Lens' (PRoot a) (Set a)
+prChildrenL = lens prChildren $ \x y -> x {prChildren = y}
+
+data PNode a = PNode
+  { pnParent :: !(Maybe a)
+  , pnChildren :: !(Set a)
+  }
+  deriving (Eq, Ord, Generic, NoThunks, NFData, Default)
+
+instance Show a => Show (PNode a) where
+  show (PNode parent children) =
+    mconcat
+      [ "<N"
+      , case parent of
+          Just p -> mconcat ["(", show p, ")"]
+          Nothing -> "(-)"
+      , ","
+      , "{"
+      , show $ Set.size children
+      , "}"
+      , show $ Set.toList children
+      , ">"
+      ]
+
+_pnParentL :: Lens' (PNode a) (Maybe a)
+_pnParentL = lens pnParent $ \x y -> x {pnParent = y}
+
+pnChildrenL :: Lens' (PNode a) (Set a)
+pnChildrenL = lens pnChildren $ \x y -> x {pnChildren = y}
+
+newtype PHierarchy a = PHierarchy
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+instance Show a => Show (PHierarchy a) where
+  show = show . Map.assocs . unPHierarchy
+
+pHierarchyNTL :: Lens' (PHierarchy a) (Map a (PNode a)) -- FIXME: what a terrible name!
+pHierarchyNTL = lens unPHierarchy $ \x y -> x {unPHierarchy = y}
```

Comment:
What is `NT`  ?

Reply:
:vomiting_face: NewType

Reply:
I want a recommendation to name this differently :yum: Can you think of a name, especially considerting that `pHierarchyL` is already taken. cc. @Soupstraw

Reply:
hmm, maybe rename `pHierarchy` to `pHierarchyForest` and then rename this to `pHierarchy`?

Reply:
Nice thanks! :raised_hands: I will use this :+1: